### DISCUSSION TITLE
Questions about  weed filer.sync

### DISCUSSION THREAD

**Discussion by guol-fnst** [seaweedfs/seaweedfs#2554] [Q&A] [answered]
weed version 2.77
ARCH:x86
I have 3 nodes on which weed were running. Every node had master,volume,filer,s3 and iam service.

I used command "weed filer.sync -a localhost:28000 -b 20.20.20.91:28000 -isActivePassive'  on **Node-1**  to sync file to another seaweedfs cluster.
Then I made some tests.
Step 1: I upload some files on **Node-2** through S3 interface.
Step 2: The new uploaded files didn't show on the other seaweedfs  cluster.
Step 3: I restart "weed filer.sync" on **Node-2**, the new uploaded files appeared on the other seaweedfs  cluster.
Howerver if I uploaded files on **Node-1**, where process "weed filer.sync" was running, filer sync worked normally.

So I guess "weed filer.sync" only monitor the the file changes of locale filer. In this case, is it ok to start "weed filer.sync" on every node to ensure all the files are  synchronized normally. 

Here are the commands on node1-node3:
Node1:
```
weed -v=0 master -mdir=/mnt/sss/object_aaaa/master_1 -port=26000 -ip=20.20.21.11 -peers=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 s3 -port=29000 -filer=20.20.21.11:28000
weed -v=0 iam -filer=20.20.21.11:28000 -master=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 volume -dir=/mnt/sss/object_aaaa/data_3 -port=27002 -ip=20.20.21.11 -max 4294967295 -mserver=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 filer -port=28000 -master=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
```
Node2:
```
weed -v=0 volume -dir=/mnt/sss/object_aaaa/data_1 -port=27000 -ip=20.20.21.12 -max 4294967295 -mserver=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 master -mdir=/mnt/sss/object_aaaa/master_2 -port=26000 -ip=20.20.21.12 -peers=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 filer -port=28000 -master=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 s3 -port=29000 -filer=20.20.21.12:28000
weed -v=0 iam -filer=20.20.21.12:28000 -master=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
```
Node3:
```
weed -v=0 master -mdir=/mnt/sss/object_aaaa/master_3 -port=26000 -ip=20.20.21.13 -peers=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 volume -dir=/mnt/sss/object_aaaa/data_2 -port=27001 -ip=20.20.21.13 -max 4294967295 -mserver=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 filer -port=28000 -master=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
weed -v=0 s3 -port=29000 -filer=20.20.21.13:28000
weed -v=0 iam -filer=20.20.21.13:28000 -master=20.20.21.11:26000,20.20.21.12:26000,20.20.21.13:26000
```

**Comment by chrislusf** [comment, marked as answer]
`weed filer.sync -a localhost:28000 -b 20.20.20.91:28000 -isActivePassive` This is one-directional replicating data from cluster A to cluster B.

But your setup is not 3 clusters, but just one cluster. Use the latest version, which you would not need to use `weed filer.sync`.

**Comment by guol-fnst** [comment]
Maybe I did not describe clearly.  
Filer "20.20.20.91:28000" belongs to cluster B.
I want to sync files from A to B using "weed filer.sync -a localhost:28000 -b 20.20.20.91:28000 -isActivePassive' on Node-1 of cluster A.

It seems thart "weed filer.sync " only sync new files  of  locale filer?

**Reply by guol-fnst** [reply]
I uploaded file to filer on node2. The new uploaded file was not synced to cluster B. Mybe "weed filer.sync" on node-1 didn't monitor the changes on  node-2.

**Reply by chrislusf** [reply]
Your command line option `-isActivePassive` is one-directional replicating data from cluster A to cluster B. So the process does not listen for cluster B updates.

**Reply by chrislusf** [reply]
> I uploaded file to filer on node2. The new uploaded file was not synced to cluster B.

I am confused. I thought node2 is in cluster B.

**Reply by guol-fnst** [reply]
So is it write about my conclusion "weed filer.sync  only sync new files of locale filer" ?
And to overcome this problem, is it ok to start "weed filer.sync" on every node to ensure all the files are synchronized normally to Cluster B.

**Reply by chrislusf** [reply]
"weed filer.sync only sync new files of locale filer" No. By default it is active-active. But you used `-isActivePassive`, so it is one directional. 

It is also not just for local filer, but for the whole cluster. But your cluster has configuration issues. You need to use the latest version.

**Reply by guol-fnst** [reply]
"weed filer.sync only sync new files of locale filer":-> My description may counfused you again. I can explain it.
Backgroud : Node-1, node-2 and node 3 belongs to cluster A. I tried to synce filer from cluster A to another weed cluster B.
 I started weed filer on node-1 ,node-2 and node-3.  And weed filer.sync were started on node-1  All the existed files were synced normally.
Then I uploaded a new file through filer on node-2. But the new file was not synced to cluster B. Only if I restarted filer.sync on node-1, then the new file could be synced to cluster B. 
I will try the latest version tomorrow.

Another thing, could you please point out the mistakes of my cluster configuration?
Very appreciate for your help!

**Reply by chrislusf** [reply]
For version 2.77, you would still need to add "-peers" for filers, to get the filers form a cluster.

For latest version, this is not needed.

**Reply by guol-fnst** [reply]
I used the latest version(2.8.3), everything worked fine.
Thanks !

**Comment by LuckyFrost** [comment]
I understand correctly that if I have three synchronized filers (With Automatic Peer Discovery in one cluster), then I have to run filer.sync three times (on all three servers), right?

**Reply by chrislusf** [reply]
One is enough for the whole cluster.

**Reply by LuckyFrost** [reply]
How then is fault tolerance achieved? One filer can die.

**Reply by chrislusf** [reply]
The `filer.sync` is an async operation and should be able to allow some down time.

**Reply by LuckyFrost** [reply]
Hmmm... Let's imagine that one filer has disappeared completely and irrevocably (burned out) - can I run filer.sync on another server and everything that was not replicated due to the death of the first filer.sync is then automatically replicated?
What about scaling? All the traffic will still have to be passed through only 1 server? If so, then it's sad.

**Reply by chrislusf** [reply]
> Hmmm... Let's imagine that one filer has disappeared completely and irrevocably (burned out) - can I run filer.sync on another server and everything that was not replicated due to the death of the first filer.sync is then automatically replicated?

Yes. It should resume from where it is left over.


> What about scaling? All the traffic will still have to be passed through only 1 server? If so, then it's sad.

This part is not worked on yet. Usually the network is the bottleneck when crossing data centers.
